remotes/fixup.go

```diff
@@ -3,86 +3,127 @@ package remotes
 	"github.com/containerd/containerd/remotes/docker"
 	"github.com/deislabs/duffle/pkg/bundle"
-	"github.com/docker/cli/opts"
 	"github.com/docker/distribution/reference"
 	ocischemav1 "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
 // FixupBundle checks that all the references are present in the referenced repository, otherwise it will mount all
 // the manifests to that repository. The bundle is then patched with the new digested references.
-func FixupBundle(ctx context.Context, b *bundle.Bundle, ref reference.Named, resolver docker.ResolverBlobMounter) error {
+func FixupBundle(ctx context.Context, b *bundle.Bundle, ref reference.Named, resolver docker.ResolverBlobMounter, events chan<- FixupEvent) error {
 	if len(b.InvocationImages) != 1 {
 		return fmt.Errorf("only one invocation image supported for bundle %q", ref)
 	}
 	var err error
-	if b.InvocationImages[0].BaseImage, err = fixupImage(ctx, b.InvocationImages[0].BaseImage, ref, resolver); err != nil {
+	if b.InvocationImages[0].BaseImage, err = fixupImage(ctx, b.InvocationImages[0].BaseImage, ref, resolver, events); err != nil {
 		return err
 	}
 	for name, original := range b.Images {
-		if original.BaseImage, err = fixupImage(ctx, original.BaseImage, ref, resolver); err != nil {
+		if original.BaseImage, err = fixupImage(ctx, original.BaseImage, ref, resolver, events); err != nil {
 			return err
 		}
 		b.Images[name] = original
 	}
 	return nil
 }
 
-func fixupImage(ctx context.Context, baseImage bundle.BaseImage, ref reference.Named, resolver docker.ResolverBlobMounter) (bundle.BaseImage, error) {
-	fmt.Fprintf(os.Stderr, "Ensuring image %s is present in repository %s\n", baseImage.Image, ref.Name())
+func fixupImage(ctx context.Context, baseImage bundle.BaseImage, ref reference.Named, resolver docker.ResolverBlobMounter, events chan<- FixupEvent) (bundle.BaseImage, error) {
+	progress := &progress{}
+	events <- FixupEvent{
+		DestinationRef: ref,
+		SourceImage:    baseImage.Image,
+		EventType:      FixupEventTypeCopyImageStart,
+		Progress:       progress.snapshot(),
+	}
+	originalSource := baseImage.Image
+	notifyEvent := func(eventType FixupEventType, message string, err error) {
+
+		events <- FixupEvent{
+			DestinationRef: ref,
+			SourceImage:    originalSource,
+			EventType:      eventType,
+			Message:        message,
+			Error:          err,
+			Progress:       progress.snapshot(),
+		}
+	}
+	notifyImageEndError := func(err error) {
+		notifyEvent(FixupEventTypeCopyImageEnd, "", err)
+	}
 	repoOnly, imageRef, descriptor, err := fixupBaseImage(ctx, &baseImage, ref, resolver)
 	if err != nil {
+		notifyImageEndError(err)
 		return bundle.BaseImage{}, err
 	}
 	if imageRef.Name() == ref.Name() {
+		events <- FixupEvent{
+			DestinationRef: ref,
+			SourceImage:    originalSource,
+			EventType:      FixupEventTypeCopyImageEnd,
+			Message:        "Nothing to do: image reference is already present in repository" + repoOnly.String(),
+		}
```

Comment:
Not using the helper and/or the population of `Progress` for End events is inconsistent.